lumicks/pylake/kymotracker/detail/peakfinding.py

```diff
@@ -0,0 +1,166 @@
+                )
+            )
+
+    def reset_assignment(self):
+        for frame in self.frames:
+            frame.reset_assignment()
+
+    def flatten(self):
+        coordinates = np.array([])
+        time_points = np.array([])
+        peak_amplitudes = np.array([])
+        for frame in self.frames:
+            coordinates = np.hstack((coordinates, frame.coordinates))
+            time_points = np.hstack((time_points, frame.time_points))
+            peak_amplitudes = np.hstack((peak_amplitudes, frame.peak_amplitudes))
+
+        return coordinates, time_points, peak_amplitudes
+
+
+def refine_peak_based_on_moment(data, coordinates, time_points, half_kernel_size, max_iter=100, eps=1e-7):
+    """This function adjusts the coordinates estimate by a brightness weighted centroid around the initial estimate.
+    This estimate is obtained by filtering the image with a kernel. If a pixel offset has a larger magnitude than 0.5
+    then the pixel is moved and the centroid recomputed. The process is repeated until there are no more changes.
+    Convergence usually occurs within a few iterations.
+
+    Parameters
+    ----------
+    data : array_like
+        A 2D image of pixel data (first axis corresponds to coordinates, second to time points).
+    coordinates : array_like
+        Initial coordinate estimates.
+    time_points : array_like
+        Time points at which the coordinate estimates were made.
+    half_kernel_size : int
+        Half of the kernel size in pixels. The kernel is used to refine the line estimate. The kernel size used for this
+        refinement will be 2 * half_kernel_size + 1.
+    max_iter : int
+        Maximum number of iterations
+    eps : float
+        We add a little offset to the normalization to prevent divisions by zeros on pixels that did not have any photon
+        counts. Eps sets this offset.
+    """
+    half_kernel_size = int(math.ceil(half_kernel_size))
+    dir_kernel = np.expand_dims(np.arange(half_kernel_size, -(half_kernel_size + 1), -1), 1)
+    mean_kernel = np.ones((2 * half_kernel_size + 1, 1))
+    coordinates = np.copy(coordinates)
+
+    m0 = convolve2d(data, mean_kernel, 'same')
+    subpixel_offset = convolve2d(data, dir_kernel, 'same') / (m0 + eps)
+
+    iteration = 0
+    done = False
+    max_coordinates = subpixel_offset.shape[0]
+    while not done:
+        offsets = subpixel_offset[coordinates, time_points]
+        out_of_bounds, = np.nonzero(abs(offsets) > 0.5)
+        coordinates[out_of_bounds] += np.sign(offsets[out_of_bounds]).astype(np.int)
+
+        # Edge cases (literally)
+        low = coordinates < 0
+        coordinates[low] = 0
+        high = coordinates >= max_coordinates
+        coordinates[high] = max_coordinates - 1
+
+        done = out_of_bounds.size - np.sum(low) - np.sum(high) == 0
+
+        if iteration > max_iter:
+            raise RuntimeError("Iteration limit exceeded")
+
+        iteration += 1
```

Comment:
👍 I like the second one, since it draws attention to the (expected) early return. It also makes it easier to quickly glance at the function and see its return argument at a first glance (the end).